proto_parser/src/proto_parser.cc

```diff
@@ -0,0 +1,249 @@
+/*
+*   Copyright 2020 Google LLC
+*
+*   Licensed under the Apache License, Version 2.0 (the "License");
+*   you may not use this file except in compliance with the License.
+*   You may obtain a copy of the License at
+*
+*    https://www.apache.org/licenses/LICENSE-2.0
+*
+*   Unless required by applicable law or agreed to in writing, software
+*   distributed under the License is distributed on an "AS IS" BASIS,
+*   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+*   See the License for the specific language governing permissions and
+*   limitations under the License.
+*/
+
+#include "proto_parser.h"
+
+namespace wireless_android_play_analytics {
+
+void ProtoParser::PopulateFields(int& prev_field_line, 
+  const google::protobuf::TextFormat::ParseInfoTree* tree,
+  const google::protobuf::Message& message,
+  std::shared_ptr<ProtoValue>& proto_value) {
+  const google::protobuf::Reflection* reflection = message.GetReflection();
+  const google::protobuf::Descriptor* descriptor = message.GetDescriptor();
+  std::shared_ptr<MessageValue> message_tmp = 
+    std::dynamic_pointer_cast<MessageValue>(proto_value);
+  assert(message_tmp);
+  std::vector<std::shared_ptr<ProtoValue>>& message_field = message_tmp->
+    GetFieldsMutable();
+  int field_count = descriptor->field_count();
+  std::vector<FieldInfo> field_info_list;
+  // Get the order of textproto
+  for (int i = 0; i < field_count; ++i) {
+    const google::protobuf::FieldDescriptor* field_descriptor = 
+      descriptor->field(i);
+    if (field_descriptor->is_repeated()) {
+      int field_size = reflection->FieldSize(message, field_descriptor);
+      for (int j = 0; j < field_size; ++j) {
+        field_info_list.push_back(FieldInfo(GetLocation(tree, field_descriptor, 
+          j), j, field_descriptor));
+      }
+    } else if(reflection->HasField(message, field_descriptor)) {
+      field_info_list.push_back(FieldInfo(GetLocation(tree, field_descriptor, 
+      -1), -1, field_descriptor));
+    }
+  }
+  std::sort(field_info_list.begin(), field_info_list.end());
+  // Get every field of this proto.
+  for (const FieldInfo& field : field_info_list) {
+    if (field.field_descriptor_->cpp_type() == 
+      google::protobuf::FieldDescriptor::CppType::CPPTYPE_MESSAGE) {
+      const google::protobuf::Message* nested_message = nullptr;
```

Comment:
const google::protobuf::Message nested_message

You shouldn't use pointer here. You will get memory leak issue in your current implementation.